ibis/pandas/execution.py

```diff
@@ -573,42 +587,74 @@ def execute_not_bool(op, data, **kwargs):
+    else:
+        name = util.guid()
+        new_column = execute(column_expr, **kwargs)
+
+    root_table, = column_op.root_tables()
+    return name, new_column, root_table
+
+
+_LEFT_JOIN_SUFFIX = '_ibis_left_{}'.format(util.guid())
+_RIGHT_JOIN_SUFFIX = '_ibis_right_{}'.format(util.guid())
+_JOIN_SUFFIXES = _LEFT_JOIN_SUFFIX, _RIGHT_JOIN_SUFFIX
+
+
 @execute_node.register(ops.Join, pd.DataFrame, pd.DataFrame)
 def execute_materialized_join(op, left, right, **kwargs):
     try:
         how = _JOIN_TYPES[type(op)]
     except KeyError:
         raise NotImplementedError('{} not supported'.format(type(op).__name__))
 
-    overlapping_columns = set(left.columns) & set(right.columns)
+    left_op = op.left.op()
+    right_op = op.right.op()
 
-    left_on = []
-    right_on = []
+    on = {left_op: [], right_op: []}
+    new_columns = {left_op: {}, right_op: {}}
 
     for predicate in map(operator.methodcaller('op'), op.predicates):
         if not isinstance(predicate, ops.Equals):
             raise TypeError(
                 'Only equality join predicates supported with pandas'
             )
-        left_name = predicate.left._name
-        right_name = predicate.right._name
-        left_on.append(left_name)
-        right_on.append(right_name)
+        left_name, new_left_column, left_pred_root = compute_join_column(
+            predicate.left,
+            **kwargs
+        )
+        assert left_pred_root in new_columns
 
-        # TODO(phillipc): Is this the correct approach? That is, can we safely
-        #                 ignore duplicate join keys?
-        overlapping_columns -= {left_name, right_name}
+        if new_left_column is not None:
+            new_columns[left_pred_root][left_name] = new_left_column
 
-    if overlapping_columns:
-        raise ValueError(
-            'left and right DataFrame columns overlap on {} in a join. '
-            'Please specify the columns you want to select from the join, '
-            'e.g., join[left.column1, right.column2, ...]'.format(
-                overlapping_columns
-            )
+        on[left_pred_root].append(left_name)
+
+        right_name, new_right_column, right_pred_root = compute_join_column(
+            predicate.right,
+            **kwargs
         )
+        assert right_pred_root in new_columns
+
+        if new_right_column is not None:
+            new_columns[right_pred_root][right_name] = new_right_column
+
+        on[right_pred_root].append(right_name)
 
-    return pd.merge(left, right, how=how, left_on=left_on, right_on=right_on)
+    left_rel = left.assign(**new_columns[left_op])
+    right_rel = right.assign(**new_columns[right_op])
+
+    to_drop = list(itertools.chain.from_iterable(new_columns.values()))
+
+    result = pd.merge(
+        left_rel, right_rel,
+        how=how, left_on=on[left_op], right_on=on[right_op],
+        suffixes=_JOIN_SUFFIXES,
+    )
+    return result.drop(to_drop, axis=1)
```

Comment:
Can you get away with not assigning new columns to `left` and `right` (and potentially avoiding some computation) and instead passing the join columns as Series arguments to `left_on` and `right_on`, or does that not work?

Reply:
Yep, that worked. Thanks for the suggestion.